python/src/iceberg/schema.py

```diff
@@ -262,10 +267,45 @@ def primitive(self, primitive: PrimitiveType) -> T:
         ...  # pragma: no cover
 
 
+@dataclass(init=True, eq=True)
+class Accessor:
+    """An accessor for a specific position in a container that implements the StructProtocol"""
+
+    position: int
+    inner: Optional["Accessor"] = None
+
+    def __str__(self):
+        return f"Accessor(position={self.position},inner={self.inner})"
+
+    def __repr__(self):
+        return self.__str__()
+
+    def with_inner(self, accessor: "Accessor") -> "Accessor":
+        parent = copy.deepcopy(self)
+        acc = parent
+        while acc.inner:
+            acc = acc.inner
+        acc.inner = accessor
+        return parent
+
+    def get(self, container: StructProtocol) -> Any:
+        """Returns the value at self.position in `container`
+
+        Args:
+            container(StructProtocol): A container to access at position `self.position`
+
+        Returns:
+            Any: The value at position `self.position` in the container
+        """
+        return container.get(self.position)
+
+
 @singledispatch
 def visit(obj, visitor: SchemaVisitor[T]) -> T:
     """A generic function for applying a schema visitor to any point within a schema
 
+    The function traverses the schema in pre-order fashion
```

Comment:
For building the Accessors, the pre-order is much easier since we can just wrap the field in the `inner` function. Also, I think it is much easier for most people to reason because it follows the pattern of a recursive function.

I was looking at the existing `_IndexByName` visitor, and it looks like we're actually using pre-order traversal as well. But we misuse the before and after hooks: https://github.com/apache/iceberg/blob/1bb526264b22372bf94a0d28ccf626da93ea6014/python/src/iceberg/schema.py#L391-L410

I could implement a similar pattern for creating the accessors. WDYT @samredai ?

Reply:
The purpose of the visitor pattern is to separate the logic that traverses the tree structure from the logic that operates on it. Moving to pre-order conflicts with that goal because you have to inspect the inner type using `instanceof`. It think it looks more like recursive method that traverses the schema because it has to redo the work of the visitor.

Iceberg uses the pattern of using a visitor to build a tree structure quite a bit, and to do that you almost always want to use post-order so that you can wrap what was returned by visiting the children of a node. There are other ways to do it, but you end up keeping state in the visitor or needing to pass state down the tree. I think it's cleaner and easier to work with using the existing pattern. But even if that's not convincing, I'm concerned about having to rewrite all of the Java visitors to use different logic, rather than being able to see what they're doing and port the same logic. That's going to be quite a bit harder.

Reply:
I agree that pre-order feels more intuitive, especially since the index gets built in ascending field ID order. That being said, it's not worth it if there are gotchas that we're missing here.

> ...and I think makes the visitor you're building here more complicated because it can't rely on traversal to create the child accessors it needs.

This is a great point--I'm noticing now that the visit method for NestedField here is becoming responsible for much than just visiting the NestedField. I couldn't think of a post-order solution for creating the wrapped accessor after the inner accessor is visited, but I completely missed what we're doing there in `_IndexByName`. I think mimicking that is a good idea and keeping something like a `self._inner` property on the `_BuildPositionAccessors` class.

Reply:
@rdblue I've seen the light, thank you. Just reverted it to post-order traversal. It took me a while to wrap my head around it, but I like the fact that there is no state in the visitor itself. This makes is much cleaner.